### DISCUSSION TITLE
Using getServerSideProps to generate a dynamic /robots.txt file in a Next.js app

### DISCUSSION THREAD

**Discussion by cbdeveloper** [vercel/next.js#24786] [Help] [answered]
I need a way to answer dynamically to the `/robots.txt` request.

Similar to https://github.com/vercel/next.js/discussions/18697 , but he was using `getInitialProps`

And that's why I've decided to go with `getServerSideProps`

https://nextjs.org/docs/basic-features/data-fetching#getserversideprops-server-side-rendering

> If you export an async function called getServerSideProps from a page, Next.js will pre-render this page on each request using the data returned by getServerSideProps.

```
export async function getServerSideProps(context) {
  return {
    props: {}, // will be passed to the page component as props
  }
}
```

Inside context we have the `req` and `res` object.

[![enter image description here][1]][1]

The response for the `robots.txt` will depend on the `req.headers.host` value.

For example:

- `www.mydomain.com` should render a production `robots.txt` file
- `test.mydomain.com` should render a test `robots.txt` file (that I'll use on test/staging deployments).

This is my current code:

**pages/robots.txt.tsx**

```
import React from "react";
import { GetServerSideProps } from "next";

interface Robots {
  ENV: "TEST" | "PROD"
}

export const getServerSideProps : GetServerSideProps<Robots> = async (context) => {
  const { req, res } = context;
  const { host } = req.headers;

  res.write("XXX");
  res.end();

  return({                // This is unnecessary (but Next.js requires it to be here)
    props: {
      ENV: "TEST"
    }
  });
};

const Robots: React.FC<Robots> = (props) => {  // This is also unnecessary (but Next.js requires it to be here)
  console.log("Rendering Robots...");

  return(
    <div>
      I am Robots 
    </div>
  );
};

export default Robots;  // This is also unnecessary (but Next.js requires it to be here).
```

It seems to be working:

[![enter image description here][2]][2]

But the weird part is that `Next.js` demands me to export a component from that page. And also returns a `props: {}` object from `getServerSideProps` is also required.

What is the way to go here? I'm basically using `req,res` from `getServerSideProps` to return something that is not a page. Is this an anti-pattern? How should I approach this?

  [1]: https://i.stack.imgur.com/10rvA.png
  [2]: https://i.stack.imgur.com/hoiLy.png

**Comment by wwwdepot** [comment]
You can use `getInitialProps` if you don't want to use `getServerSideProps`

```jsx
// pages/robots.txt.js
class Robots extends React.Component {
  static async getInitialProps({ res }) {
    // generate robots.txt content
    res.setHeader('Content-Type', 'text/plain');
    res.write('...robots.txt content');
    res.end();
  }
}

export default Robots ;
```

**Reply by bobarros** [reply]
I have the feeling that it's not that he doesn't want to use getServerSideProps. He just wants to have a blank page.
Maybe he would be happy using it, as long as the same result it's possible.

```jsx
const Blank = () => {}
export default Blank;
export async function getServerSideProps(context) {
   context.res.setHeader('Content-Type', 'text/plain');
   context.res.write('...robots.txt content');
   context.res.end();
}
```

**Reply by cbdeveloper** [reply]
Yeah, I don't want to use `getInitialProps`. From the docs below, it seems it's not recommended for newer versions. And I'm using Next v10.

From: https://nextjs.org/docs/api-reference/data-fetching/getInitialProps

About the page, it's not that I want a blank page. It feels weird to add a page in the first place, because it seems that that page is never rendered (and it shouldn't be). Because I'm responding the request with `res.write()`. The fact that the `Page` component is required makes my code feel "hacky" and like an anti-pattern.

My question is: is this a bad practice? Is there a better solution to this? What should be the recommended approach here?

**Comment by cbdeveloper** [comment, marked as answer]
A regular page with `getServerSideProps` is not the right pattern for this job, even though it seems to be working.

The proper way of doing it is by using `rewrites` on `next.config.js` and handle it in an `pages/api` request.

https://nextjs.org/docs/api-reference/next.config.js/rewrites

**next.config.js**

```
module.exports = {

async rewrites() {
    return [
      {
        source: "/robots.txt",
        destination: "/api/robots"
      },
      {
        source: "/sitemap.xml",
        destination: "/api/sitemap"
      }
    ]
  }

};
```


**pages/api/robots.ts**

```
import { NextApiHandler } from "next";

const robotsHandler: NextApiHandler = async (req, res) => {
  
  const robotsTxt = ["User-agent: *"];
  robotsTxt.push("Disallow: /");

  res.setHeader("Cache-Control","private");
  res.setHeader("Content-Type","text/plain");
  
  return res.send(robotsTxt.join("\n"));
};


export default robotsHandler;
```
